OpenCL_C.txt

```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
+
+If the generic address space is supported i.e. for OpenCL C 2.0 or OpenCL C 3.0
+with `__opencl_c_generic_address+space` feature, pointers that are declared
+without pointing to a named address space point to the generic address space.
+
+All string literal storage shall be in the `+__constant+` address space.
+
+For all other cases that are not listed above the address space is deduced to
+private. This includes:
+
+  * All function arguments as well as return values are in private address
+    space.
+
+  * Pointers that are declared without pointing to a named address space point
+    to the `+__private+` address space if generic address space is not
+    supported.
+
+  * Variables inside a function not declared with an address space qualifier
+    are deduced to private address space.
+
+Examples:
+ 
+[source,c]
+----------
+// Note: these examples assume OpenCL C 2.0 or the
+// __opencl_c_program_scope_global_variables feature macro.
+
+int foo;                // Declared in the global address space.
+
+static int foo;         // Declared in the global address space.
+
+int *ptr;               // ptr is allocated in the global address space.
+                        // ptr points to a location in (1) the generic address
+                        // space for OpenC 2.0 or OpenCL C 3.0 with
+                        // __opencl_c_generic_address_space feature or
+                        // in (2) the private address space otherwise.
+
+int * global ptr;       // ptr points to an location in (1) the generic address
+                        // space for OpenC 2.0 or OpenCL C 3.0 with
+                        // __opencl_c_generic_address_space feature or
+                        // in (2) the private address space otherwise.
+
+constant int *ptr =
+               "Hello"; // string literal is in constant address space.
+
+void func(int param)    // param is allocated in the private address space.
+{
+    int foo;            // foo is allocated in the private address space.
+    static int foo;     // foo is allocated in the global address space.
+    int *ptr;           // ptr is allocated in the private address space.
+                        // ptr points to a location in (1) the generic address
+                        // space for OpenC 2.0 or OpenCL C 3.0 with
+                        // __opencl_c_generic_address_space feature or
+                        // in (2) the private address space otherwise.
+    ...
+}
+----------
+
+[NOTE]
+====
+Qualifiers must be explicitly specified for:
+
+  * Program scope variables or variables inside functions with
+    `static`/`extern` type specifier for OpenCL C prior to version 2.0 or
+    OpenCL C 3.0 without `+__opencl_c_program_scope_global_variables+` feature, 
+
+  * In pointers used as arguments to the kernel function (address space pointed
+    to must be speficied explicitly).
+====
+
+[[table-addr-spaces-summary]]
+.Address space behavior
+[%header,cols=4*]
+|====
```

Comment:
I like the content of this table but I think it needs to be cleaned up a bit editorially.  I can help with this if you'd like, but a few items to consider:

1. The table cells are awkwardly sized, especially in the HTML spec.  I think this is because the HTML spec doesn't want to break the feature names.  We should consider defining asciidoc attributes with zero-width spaces to help the feature names break properly.

2. Asciidoc doesn't consider the `-` character to create bulleted lists.  I think the table would look a bit tidier with proper lists, used in all cells consistently, or removing all bulleted lists entirely.

3. I think "Addr space" is over-abbreviated.  Any reason not to spell out "Address Space"?

4. Consider using the formatting e.g. "`__global`" in the left-most column vs. "Global".

Reply:
I have addressed 2-4, but I am not sure about 1. I have found this https://stackoverflow.com/questions/40464460/how-to-do-word-break-in-ascii-doc telling that the functionality not supported. Perhaps the reason for it is that in browsers it is possible to scroll horizontally?